torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -226,6 +185,11 @@ def __init__(
         self.rank = dist.get_rank(self.process_group)
         self.global_rank = _get_global_rank(self.process_group, self.rank)
 
+        # Set the current device if using NCCL backend; used for `broadcast()`
+        if dist.get_backend(self.process_group) == dist.Backend.NCCL and \
+                self._default_device.type == "cuda":
+            torch.cuda.set_device(self._default_device)
```

Comment:
will this leads to silent side effects, i.e., changing global current device?

Reply:
Great catch. I have removed this logic in favor of using the `torch.cuda.device()` context manager for the broadcasts (with the argument `self._default_device`). Though, it is a bit hacky since `self._default_device` could be CPU.